src/expiring_hash_map/mod.rs

```diff
@@ -0,0 +1,200 @@
+    ///    We were choosing between, effectively, the current "drain"-like API,
+    ///    and the "queue" like API.
+    ///    Current ("drain"-like) API waits on the deadline or returns `None`
+    ///    when there are no more items. Very similar how we [`Vec::drain`] iter
+    ///    works.
+    ///    The "queue"-like API would, pretty much, be simply waiting expired
+    ///    items to appear. In the case of empty [`ExpiringHashMap`], we would
+    ///    wait indefinetly - or until an item is inserted. This would be
+    ///    possible to carry on, for instance, from a sibling branch of a
+    ///    `select` statement, so the borrowing rules won't be a problem here.
+    /// 3. We went over the following alternative signature:
+    ///    ```ignore
+    ///    pub fn next_expired(&mut self) -> Option<impl Future<Outcome = Result<ExpiredItem<K, V>, Error>>> {...}
+    ///    ```
+    ///    This captures the API restrictions a bit better, and should provide
+    ///    less possibilities to misuse the API.
+    ///    We didn't pick this one becuase it's not an `async fn` and we wanted
+    ///    this, see (1) of this list. Furtherore, instead of doing a
+    ///    `select { _ = map.next_expired(), if !map.is_empty() => { ... } }`
+    ///    users would have to do
+    ///    `let exp = map.next_expired(); select { _ = exp.unwrap(), if exp.is_some() => { ... } }`,
+    ///    which is less readable and a bit harder to understand. Although it
+    ///    has a possiblity of a nicer generalization if `select` macro
+    ///    supported a `Some(future)` kind of pattern matching, we decided to go
+    ///    with other solution for now.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use tokio02 as tokio;
+    /// # let mut rt = tokio::runtime::Runtime::new().unwrap();
+    /// # rt.block_on(async {
+    /// use vector::expiring_hash_map::ExpiringHashMap;
+    /// use std::time::Duration;
+    ///
+    /// let mut map: ExpiringHashMap<String, String> = ExpiringHashMap::new();
+    ///
+    /// loop {
+    ///     tokio::select! {
+    ///         // Avoid spin lock!
+    ///         val = map.next_expired(), if !map.is_empty() => match val {
+    ///             None => unreachable!(), // we never poll the empty map in the first place!
+    ///             Some(Ok((val, _))) => {
+    ///                 println!("Expired: {}", val);
+    ///                 break;
+    ///             }
+    ///             Some(Err(err)) => panic!(format!("Timer error: {:?}", err)),
+    ///         },
+    ///         _ = tokio::time::delay_for(Duration::from_millis(100)) => map.insert(
+    ///             "key".to_owned(),
+    ///             "val".to_owned(),
+    ///             Duration::from_millis(30),
+    ///         ),
+    ///     }
+    /// }
+    /// # });
+    /// ```
+    pub async fn next_expired(&mut self) -> Option<Result<ExpiredItem<K, V>, Error>> {
+        let key = self.expiration_queue.next().await?;
```

Comment:
Yes, it's a https://docs.rs/tokio/0.2.13/tokio/time/struct.Error.html

Reply:
Makes sense. This is a little bit of a bummer to expose since it's something we'd usually unwrap, but I'm fine with either as long as we make the nature of the possible error clear.

Reply:
Right! I forgot to add the errors section :frowning:

Reply:
huh this is odd I thought we removed errors for timers from tokio 0.2 :thinking:

Reply:
Do timers just panic now?
It might make sense to return errors from timers IMO, but I'm not aware of how the development unfolds in `tokio`. :smile:  Either way - it doesn't make sense to go ahead of the `tokio` changes. We'll just remove errors here if/when relevant `tokio` update arrives.